internal/internal_eager_activity.go

```diff
@@ -0,0 +1,157 @@
+	req *workflowservice.RespondWorkflowTaskCompletedRequest,
+) (amountReserved int) {
+	// Go over every command checking for activities that can be eagerly executed
+	for _, command := range req.Commands {
+		if attrs := command.GetScheduleActivityTaskCommandAttributes(); attrs != nil {
+			// If not present, disabled, or on a different task queue, we must mark as
+			// explicitly disabled
+			if e == nil || e.disabled || e.activityWorker == nil || e.taskQueue != attrs.TaskQueue.GetName() {
+				attrs.RequestEagerExecution = false
+			} else if attrs.RequestEagerExecution {
+				// If it has been requested, attempt to reserve one pending
+				attrs.RequestEagerExecution = e.reserveOnePendingSlot()
+				if attrs.RequestEagerExecution {
+					amountReserved++
+				}
+			}
+		}
+	}
+	return
+}
+
+func (e *eagerActivityExecutor) reserveOnePendingSlot() bool {
+	// Lock during count checks. Nothing in here blocks including the channel
+	// receive to serve a slot.
+	e.countLock.Lock()
+	defer e.countLock.Unlock()
+	// Confirm that, if we have a max, pending + executing isn't already there
+	if e.maxConcurrent > 0 && e.executingCount+e.pendingCount >= e.maxConcurrent {
+		// No more room
+		return false
+	}
+	// Reserve a spot for our request via a non-blocking attempt to take a poller
+	// request entry which essentially reserves a spot
+	select {
+	case <-e.activityWorker.worker.pollerRequestCh:
+	default:
+		return false
+	}
+
+	// We can request, so increase the pending count
+	e.pendingCount++
+	return true
+}
+
+func (e *eagerActivityExecutor) handleResponse(
+	resp *workflowservice.RespondWorkflowTaskCompletedResponse,
+	amountReserved int,
+) {
+	// Ignore disabled or none present
+	if e == nil || e.activityWorker == nil || e.disabled || (len(resp.ActivityTasks) == 0 && amountReserved == 0) {
+		return
+	} else if len(resp.ActivityTasks) > amountReserved {
+		panic(fmt.Sprintf("Unexpectedly received %v eager activities though we only requested %v",
+			len(resp.ActivityTasks), amountReserved))
+	}
+
+	// Update counts under lock
+	e.countLock.Lock()
+	// Record the number of unfulfilled slots we have to give back
+	unfulfilledSlots := amountReserved - len(resp.ActivityTasks)
+	// Remove all pending from worker-scope pending
+	e.pendingCount -= amountReserved
+	// Add the activity count to the executing count
+	e.executingCount += len(resp.ActivityTasks)
+	e.countLock.Unlock()
+
+	// Put every unfulfilled slot back on the poller channel
+	for i := 0; i < unfulfilledSlots; i++ {
+		// Like other parts that push onto this channel, we assume there is room
+		// because we took it, so we do a blocking send
+		e.activityWorker.worker.pollerRequestCh <- struct{}{}
+	}
+
+	// Start each activity asynchronously
+	for _, activity := range resp.ActivityTasks {
+		// Before starting the goroutine we have to increase the wait group counter
+		// that the poller would have otherwise increased
+		e.activityWorker.worker.stopWG.Add(1)
```

Comment:
IMO encapsulate this behind a function in activity worker

Reply:
Agreed line count matters little (most things can be one line if you try hard enough), it's just that this is internal code with internal knowledge of its siblings, there's little value in abstraction. I'll go ahead and comment in the other spot where the `Add` is done on poll response that this is done in eager activity also.

Reply:
Learned it's actually in 3 other places. Basically `stopWG` is used throughout to indicate "process that must complete before stop completes". So it's a known shared piece.

Reply:
Cool, IMO that's worth encapsulating with a clear name and comment, but up to you.